Add seasonal rank track to milestones

melanieseltzer opened the pull request:
Closes #4296

It's a _little_ tight in the space with the premium track, so not sure if it should be broken out into a larger space. But for now I think it's a decent stop gap. A little iffy on the white border I've added as well but figured we could address design here in the review.

With premium track:
<img width="575" alt="Screen Shot 2019-10-06 at 10 58 43 PM" src="https://user-images.githubusercontent.com/17421347/66288172-09905f00-e88d-11e9-800f-8203b6d91186.png">
<img width="379" alt="Screen Shot 2019-10-06 at 10 58 49 PM" src="https://user-images.githubusercontent.com/17421347/66288174-09905f00-e88d-11e9-899d-fc0782a73d41.png">

Without premium track:
<img width="575" alt="Screen Shot 2019-10-06 at 10 59 13 PM" src="https://user-images.githubusercontent.com/17421347/66288181-10b76d00-e88d-11e9-9dd3-b53323d179ab.png">
<img width="384" alt="Screen Shot 2019-10-06 at 10 59 05 PM" src="https://user-images.githubusercontent.com/17421347/66288180-10b76d00-e88d-11e9-804e-6c22dfe653cd.png">

bhollis commented on src/app/progress/SeasonalRank.tsx:
```diff
@@ -0,0 +1,105 @@
+  season: DestinySeasonDefinition | undefined;
+  profileInfo: DestinyProfileResponse;
+}) {
+  const formatter = new Intl.NumberFormat(window.navigator.language);
```
You can use the `numberFormatter` from `util`:

```
numberFormatter(settings.language)
```

bhollis commented on src/app/progress/SeasonalRank.tsx:
```diff
@@ -0,0 +1,105 @@
+  const formatter = new Intl.NumberFormat(window.navigator.language);
+
+  // Get season details
+  const seasonNameDisplay = season && season.displayProperties.name;
```
You might just want to return null if there's no season.

bhollis commented on src/app/progress/SeasonalRank.tsx:
```diff
@@ -0,0 +1,105 @@
+  const seasonNameDisplay = season && season.displayProperties.name;
+  const seasonPassProgressionHash = season && season.seasonPassProgressionHash;
+  const seasonEnd = season && season.endDate;
+  const { seasonHashes } = profileInfo.profile.data!;
```
Unfortunately this data can be missing - we have to check that it's present. 

Personal preference, but I tend to think there's no need to destructure a single item.

bhollis commented on src/app/progress/SeasonalRank.tsx:
```diff
@@ -0,0 +1,105 @@
+
+  return (
+    <div
+      className={`seasonal-rank milestone-quest ${hasPremiumRewards ? 'has-premium-rewards' : ''}`}
```
You can use `classNames` for this sort of thing:

```
className={classNames('seasonal-rank', 'milestone-quest', { 'has-premium-rewards': hasPremiumRewards })}
```

bhollis commented on src/app/progress/SeasonalRank.tsx:
```diff
@@ -0,0 +1,105 @@
+            </div>
+          );
+        })}
+        <span>
```
We should copy in the progress bar from PursuitItem as well.

bhollis commented on src/app/progress/SeasonalRank.tsx:
```diff
@@ -0,0 +1,105 @@
+                title={itemInfo.displayProperties.description}
+              />
+              {item.quantity && item.quantity > 1 && (
+                <div className="seasonal-reward-quantity">{item.quantity}</div>
```
Check out PursuitItem - it has an affordance for indicating amount of an item.

bhollis commented on src/app/progress/SeasonalRank.tsx:
```diff
@@ -0,0 +1,105 @@
+        </span>
+      </div>
+      <div className="milestone-info">
+        <span className="milestone-name">Rank {seasonalRank}</span>
```
We'll need to translate these - check out how we use the `t` function to provide localized strings that live in `i18n.json`. It's too bad this string isn't in the manifest anywhere.

Maybe this title should be "Seasonal Rank" just to be clear.

bhollis commented on src/app/progress/milestone.scss:
```diff
@@ -146,6 +146,40 @@
+    max-width: calc(var(--item-size) * 1.2);
+  }
+
+  &.has-premium-rewards {
```
One idea would be to display the non-premium reward as a smaller overlay over the premium reward, rather than displaying them side-by-side.

bhollis commented on src/app/progress/milestone.scss:
```diff
@@ -146,6 +146,40 @@
+
+.seasonal-rank {
+  .milestone-icon {
+    max-width: calc(var(--item-size) * 1.2);
```
Why does this need to be bigger than --item-size?

bhollis commented on src/app/progress/milestone.scss:
```diff
@@ -146,6 +146,40 @@
+  .season-end {
+    font-style: italic;
+    .season-end-title {
+      text-transform: uppercase;
```
This ends up a bit shouty - maybe we could stick with the regular casing.

bhollis commented on src/app/progress/milestone.scss:
```diff
@@ -146,6 +146,40 @@
+  }
+
+  .seasonal-reward-wrapper {
+    border: 2px solid #fff;
```
For items we use a 1px border (actually `$item-border-width` in SCSS). Since some of these are transparent we should probably also give it a background color.

Check out PursuitImage for some styles that will match up with our existing pursuits display.

delphiactual commented on src/locale/dim.json:
```diff
@@ -590,6 +590,8 @@
     "Daily": "Daily Challenge",
     "OneTime": "One Time Challenge",
+    "SeasonalRank": "Seasonal Rank",
+    "SeasonEnds": "Season ends: ",
```
```suggestion
    "SeasonEnds": "Season ends: {{date}}",
```

melanieseltzer replied:
I don't think this is going to work with the countdown component as the date, unless there's a different way to do this?

```js
{t('Milestone.SeasonEnds', {
  date: <Countdown endTime={new Date(seasonEnd!)} compact={true} />
})}
```

Output

```
Season ends: [object Object]
```

delphiactual replied:
That is a shame.

melanieseltzer commented:
@bhollis I believe I've addressed most of your comments! Let me know if there's any more changes. I've updated the screenshots in my PR if you need to see it at a glance.

RE: the suggestion to display the free reward as a small overlay, I played around with it a little but to me it looked a little awkward. I feel like if we were to do this we'd have to bump up the size of the premium icon, otherwise the free icon kinda covers most of it if it's overlaid. I can post a screenshot of it if you want to see?

One other thing, do you think the progress bar should become its own component now that it's getting reused?
